examples/keyset.rs

```diff
@@ -0,0 +1,524 @@
+        return;
+    }
+    if command == "cache-expired1" {
+        do_cache_expired1(filename, &args[3..]);
+        return;
+    }
+    if command == "propagation2-complete" {
+        do_propagation2(filename, &args[3..]);
+        return;
+    }
+    if command == "cache-expired2" {
+        do_cache_expired2(filename, &args[3..]);
+        return;
+    }
+    if command == "done" {
+        do_done(filename, &args[3..]);
+        return;
+    }
+    if command == "actions" {
+        do_actions(filename, &args[3..]);
+        return;
+    }
+    if command == "status" {
+        do_status(filename, &args[3..]);
+        return;
+    }
+
+    eprintln!("Unknown command '{command}'. Valid commands are: init, addkey, deletekey, start, propagation1-complete, cache-expired1, propagation2-complete, cache-expired2, done, actions, and status");
+    exit(1);
+}
+
+fn do_init(filename: &str, args: &[String]) {
+    if args.len() != 1 {
+        eprintln!("Usage: ... init <domain>");
+        exit(1);
+    }
+    let domainname = &args[0];
+
+    let ks = KeySet::new(Name::from_str(domainname).unwrap());
+
+    save_keyset(filename, &ks);
+}
+
+fn do_addkey(filename: &str, args: &[String]) {
+    if args.len() < 2 || args.len() > 3 {
+        eprintln!("Usage: ... addkey <key-type> <pubref> [<privref>]");
+        exit(1);
+    }
+    let keytype = &args[0];
+    let pubref = args[1].clone();
+    let privref = if args.len() == 3 {
+        Some(args[2].clone())
+    } else {
+        None
+    };
+
+    let mut ks = load_keyset(filename);
+    if keytype == "ksk" {
+        ks.add_key_ksk(pubref, privref, UnixTime::now());
+    } else if keytype == "zsk" {
+        ks.add_key_zsk(pubref, privref, UnixTime::now());
+    } else if keytype == "csk" {
+        ks.add_key_csk(pubref, privref, UnixTime::now());
+    } else {
+        eprintln!("Unknown key type '{keytype}'");
+        exit(1);
+    }
+    save_keyset(filename, &ks);
+}
+
+fn do_deletekey(filename: &str, args: &[String]) {
+    if args.len() != 1 {
+        eprintln!("Usage: ... deletekey <pubref>");
+        exit(1);
+    }
+    let pubref = &args[0];
+
+    let mut ks = load_keyset(filename);
+    if let Err(e) = ks.delete_key(pubref) {
+        eprintln!("Unable to delete key {pubref}: {e}");
+        exit(1);
+    }
+    save_keyset(filename, &ks);
+}
+
+fn do_start(filename: &str, args: &[String]) {
+    if args.len() != 1 {
+        eprintln!("Usage: ... start <roll-type>");
+        exit(1);
+    }
+    let rolltype = &args[0];
+
+    let rolltype = str_to_rolltype(rolltype);
+
+    let mut ks = load_keyset(filename);
+
+    // Find old and new keys.
+    let keys = ks.keys().to_vec();
+    let (old, new): (Vec<_>, Vec<_>) = keys
+        .into_iter()
+        .filter_map(|k: Key| match rolltype {
+            RollType::KskRoll => {
+                if let KeyType::Ksk(keystate) = k.keytype() {
+                    Some((keystate.clone(), k.pubref().to_string()))
```

Comment:
Use `keys.iter()` instead of `keys.into_iter()` above, then `.to_string()` is unnecessary here.

Reply:
Done